### DISCUSSION TITLE
Alpine optimistic update out of sync

### DISCUSSION THREAD

**Discussion by spiffytech** [bigskysoftware/htmx#1367] [Q&A]
I set up a toy scenario to test out optimistic updates using Alpine.js, but it's not working. Am I doing something wrong, or is there a bug?

What I want to happen:

1. User clicks the button
2. Alpine displays the new state it predicts the server will display
3. Simultaneously, HTMX submits the request to the server
4. When the request finishes, both displayed values (client and server state) are in sync, and Alpine's state is updated to whatever the server _actually_ returned (even if it's different from predicted).

What's actually happening:

1. The first click works correctly.
5. The second click does an optimistic update, but when the request completes, both displayed values change (wrong), and they display mismatched values (`true`+`false`, `false`+`true`)
6. All subsequent clicks don't trigger Alpine's onclick logic at all, and HTMX page updates cycle the displayed values between mismatched value pairs.

```html
<!DOCTYPE html>
<html>
<body>

<div hx-target="this" hx-ext="alpine-morph" hx-swap="morph">
	<div x-data="{bool: ${state.bool}}">
		<button type="button" x-on:click="bool = !bool" hx-post="/toggle" hx-trigger="click delay:1s">
			Toggle!
		</button>
               <!-- Server state -->
		<span>${state.bool}</span>
               <!-- Client state -->
		<span x-text="bool"></span>
	</div>
</div>

<script src="https://unpkg.com/htmx.org@latest"></script>
<script src="https://unpkg.com/htmx.org/dist/ext/alpine-morph.js"></script>
<!-- Alpine Plugins -->
<script src="https://unpkg.com/@alpinejs/morph@latest/dist/cdn.min.js"></script>
<!-- Alpine Core -->
<script src="https://unpkg.com/alpinejs@latest/dist/cdn.min.js"></script>

</body>
</html>
```

**Comment by jreviews** [comment]
For this particular example, if you are updating the Alpine x-data state directly from the server and you are swapping the entire Alpine component as well, then you should try this without using the alpine morph extension to see if that works. Alpine morph was initially developed, by me, but I never used it, to keep client side state even after swapping the html. For example, if you are using tabs, to retain the selected tab state after a swap without having to keep the tab state on the server as well.

**Reply by spiffytech** [reply]
Ahhh, I see.

Updating the example to not use morphing fixes the Alpine-displayed data:

```html
<div id="wrapper" hx-target="this" hx-select="#wrapper">
	<button type="button" hx-get="">Outside Alpine</button>

	<div x-data="{bool: false}">
		<button type="button" x-on:click="bool = !bool" hx-post="/toggle" hx-trigger="click delay:1s">
			Inside Alpine
		</button>
		<!-- Server-side state -->
		<span>${state.bool}</span>
		<!-- Client-side state -->
		<span x-text="bool">empty</span>
	</div>
</div>
```

However, I do need parts of my UI to morph, and in a complex-enough UI I expect to have Alpine components within an morph context.

I'm now testing Alpine with Idiomorph (since that sounds like the new official morph solution), and I can't figure out how to make them play nice together.

In this example, clicking the `Inside Alpine` button works correctly, while clicking the `Outside Alpine` button erases Alpine's modifications to the DOM:

```html
<div id="wrapper" hx-target="this" hx-select="#wrapper" hx-ext="morph" hx-swap="morph:outerHTML">
	<button type="button" hx-get="">Outside Alpine</button>
	 
	<div x-data="{bool: false}" hx-swap="outerHTML">
		<button type="button" x-on:click="bool = !bool" hx-post="/toggle" hx-trigger="click delay:1s">
			Inside Alpine
		</button>
		<!-- Server-side state -->
		<span>${state.bool}</span>
		<!-- Client-side state -->
		<span x-text="bool">empty</span>
	</div>
</div>
```

Since the docs for `hx-swap="outerHTML"` say "replace", I figured the Alpine DOM node would get deleted + recreated, and Alpine would reinitialize it. But instead it looks like it's being preserved, so HTMX morphis it and it's left with the `empty` content.

What's the recommended approach for mix-n-matching morphing with Alpine?
